opty/tests/test_utils.py

```diff
@@ -200,6 +200,43 @@ def test_invalid_integration_limits(self):
                 sym.Integral(self.x ** 2, (self.t, 0, 1)), self.state_symbols,
                 self.input_symbols, self.unknown_symbols, self.N, 1)
 
+    def test_variable_time(self):
+
+        def expected_obj(free):
+            f = free[2*self.N:-1]
+            return free[-1]*np.sum(f**2)
+
+        def expected_obj_grad(free):
+            f = free[2*self.N:-1]
+            grad = np.zeros_like(free)
+            grad[2*self.N:-1] = 2.0*free[-1]*free[2*self.N:-1]
+            grad[-1] = np.sum(f**2)
+            return grad
```

Comment:
You should take into account that this is backward Euler, so the first term falls out, see `test_backward_single_input`.
If free is [x(t), v(t), f1(t), f2(t), c, k, m, h] then the objective should be `(f1_vals[1:]**2 + f2_vals[1:]**2).sum() * h_val`.
Similarly, the gradient should be a stack of zeros(2*N+1), 2*h_val*f1_vals[1:], [0] 2*h_val*f2_vals[1:], [0, 0, 0, (f1_vals[1:]**2 + f2_vals[1:]**2).sum()]

P.S. quickly wrote out the equations on my phone so would advise checking them.

Reply:
I don't think I assumed any specific integration routine in the manually created objective functions.